BCM2835 ISP driver

naushir opened the pull request:
Integration of the new bcm2835 ISP V4L2 driver.

naushir committed:
configs: Add VIDEO_ISP_BCM2835 to all Pi defconfigs

This commit enables the bcm2835-isp driver in the following
defconfigs:

bcm2709_defconfig
bcm2711_defconfig
bcmrpi_defconfig
bcmrpi3_defconfig

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: uapi: v4l-ctrls: Add CID base for the bcm2835-isp driver

We are reserving controls for the new bcm2835-isp driver.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
staging: vchiq: Load bcm2835_isp driver from vchiq

bcmn2835_isp is a platform driver dependent on vchiq,
therefore add the load/unload functions for it to vchiq.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir commented:
@6by9 here is the first draft of the ISP driver.  When you have the stomach for it, a first round of review comments would be appreciated.

6by9 commented on arch/arm64/configs/bcmrpi3_defconfig:
```diff
@@ -1117,6 +1117,7 @@ CONFIG_FB_TFT_WATTEROTT=m
 CONFIG_SND_BCM2835=m
 CONFIG_VIDEO_BCM2835=m
 CONFIG_VIDEO_CODEC_BCM2835=m
+CONFIG_VIDEO_ISP_BCM2835=m
```
arch/arm64/configs/bcm2711_defconfig too.

6by9 commented on drivers/media/v4l2-core/v4l2-ioctl.c:
```diff
@@ -1333,6 +1333,7 @@ static void v4l_fill_fmtdesc(struct v4l2_fmtdesc *fmt)
 	case V4L2_META_FMT_UVC:		descr = "UVC Payload Header Metadata"; break;
 	case V4L2_META_FMT_D4XX:	descr = "Intel D4xx UVC Metadata"; break;
 	case V4L2_META_FMT_SENSOR_DATA:	descr = "Sensor Ancillary Metadata"; break;
+	case V4L2_META_FMT_BCM2835_ISP_STATS: descr = "BCM2835 ISP Image Statistics"; break;
```
Do we need to make it _BCM283X_ as it applies to 2835/2836/2837, and of course the anomalous 2711?

We do have the inconsistency there in that the drivers are bcm2835-camera, bcm2835-codec, bcm2835-unicam, but I can see the user comments otherwise.

naushir replied:
Good point.  I suppose you could argue that the HW in 2836/2837/2711 has not changed so the BCM2835 name is still "right".

If we want to be consistent with the other drivers, perhaps we keep it as-is?

6by9 replied:
Sorry, forgot to reply. Leave as is.

6by9 reviewed (commented):
Quick review of the 4 simple commits. I'll try to do the big one later.

6by9 commented on Documentation/media/uapi/v4l/meta-formats.rst:
```diff
@@ -19,6 +19,7 @@ These formats are used for the :ref:`metadata` interface only.
 .. toctree::
     :maxdepth: 1
 
+    pixfmt-meta-bcm2835-isp-stats
```
Should be in the patch adding the image format.

naushir replied:
The reason I did not add the documentation to that earlier patch was because the documentation pulls structures from the isp uapi header file - which gets added in the large patch.  Perhaps I should add the uapi header file as a separate first patch?

6by9 replied:
Not going to fuss on this one for now. Upstream may, but sort that out later.

6by9 commented on Documentation/media/uapi/v4l/pixfmt-meta-bcm2835-isp-stats.rst:
```diff
@@ -0,0 +1,41 @@
+.. Permission is granted to copy, distribute and/or modify this
```
Should be in the patch adding the image format.

naushir replied:
Done.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+The BCM2835 Image Sensor Pipeline (ISP) is a fixed function hardware pipeline
+for performing image processing operations. Input frames (in Bayer format) may
+be obtained from the BCM2835 Unicam peripheral, written to a memory framebuffer,
+and fed into the input of the ISP.  Two output images may be generated (in
```
Do we not accept YUV/RGB in this implementation? Needs stating that you'll get no stats, but not reason not to allow for the dual resolution output.
Bayer image doesn't have to come from Unicam either.

naushir replied:
I worded this incorrectly, I meant to say that Bayer images could be obtained from unicam, written to memory and passed into the ISP as an example of usage.  I will reword to make things clearer.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+        :ref:`RAW12 <V4L2-PIX-FMT-SRGGB12>`,
+        :ref:`RAW12P <V4L2-PIX-FMT-SRGGB12P>`,
+        :ref:`RAW14 <V4L2-PIX-FMT-SRGGB14P>`,
+        :ref:`RAW16 <V4L2-PIX-FMT-SRGGB16>`.
```
(Sorry, reading as I go along, so haven't checked the code as yet).
Do we really support the unpacked formats? V4L2-PIX-FMT-SRGGB10 and V4L2-PIX-FMT-SRGGB12. I guess it's only a modification to the the black level block settings? Who sets that up?

naushir replied:
You are right we don't support it currently.  I will remove.

If we do need support, it is trivial.  The ISP HW will handle everything automatically (given the correct input format is selected) since Unicam will not shift the pixels up, only pack into 16-bits.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+      - source
+      - High resolution YUV or RGB processed output from the ISP.
+      - :ref:`YUV420 <V4L2-PIX-FMT-YUV420>`,
+        :ref:`NV12/NV12 <V4L2-PIX-FMT-NV12>`,
```
NV12/NV12?
NV21 possibly, but that would be a V4L2-PIX-FMT-NV21.

naushir replied:
Yes, typo on the NV12->NV21.

Note that <V4L2-PIX-FMT-NV12> is a reference tag that lands on the heading with both V4L2_PIX_FMT_NV12 (‘NV12’) and V4L2_PIX_FMT_NV21 (‘NV21’) formats.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+      - Low resolution YUV or RGB processed output from the ISP. The output of
+        this pad cannot have a resolution larger than the "bcm2835-isp0-capture1" pad in any dimension.
+      - :ref:`YUV420 <V4L2-PIX-FMT-YUV420>`,
+        :ref:`NV12/NV12 <V4L2-PIX-FMT-NV12>`,
```
Ditto,

naushir replied:
Done.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+    * - "bcm2835-isp0-output0"
+      - sink
+      - Accepts Bayer format frame buffers as input to the ISP HW pipeline.
+      - :ref:`RAW8 <V4L2-PIX-FMT-SRGGB8>`,
```
Is the ref correct in using hyphens? The types are using underscores.

naushir replied:
Yes, the ref labels do not follow the format label.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+    * - "bcm2835-isp0-capture1"
+      - source
+      - High resolution YUV or RGB processed output from the ISP.
+      - :ref:`YUV420 <V4L2-PIX-FMT-YUV420>`,
```
YVU420 as well?

naushir replied:
Done.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+      - source
+      - Low resolution YUV or RGB processed output from the ISP. The output of
+        this pad cannot have a resolution larger than the "bcm2835-isp0-capture1" pad in any dimension.
+      - :ref:`YUV420 <V4L2-PIX-FMT-YUV420>`,
```
Ditto YVU420.

naushir replied:
Done.

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+
+    * - "bcm2835-isp0-capture2"
+      - source
+      - Low resolution YUV or RGB processed output from the ISP. The output of
```
YUV only, not RGB.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/Kconfig:
```diff
@@ -0,0 +1,11 @@
+	help
+	  Say Y here to enable the V4L2 ISP driver for
+	  Broadcom BCM2835 SoC. This operates over the VCHIQ interface
+	  to a service running on VideoCore.
```
Normally includes the resulting module name as well.

naushir replied:
Could you point to a template or example for this?  I cannot seem to find any to follow?

6by9 replied:
It seems to vary whether they do or not., so obviously not too critical
https://github.com/raspberrypi/linux/blob/rpi-5.4.y/drivers/media/platform/Kconfig#L109 is an example, as are numerous of the other drivers in that Kconfig.

6by9 commented on drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h:
```diff
@@ -23,21 +23,21 @@
 
 /** Common parameter ID group, used with many types of component. */
-#define MMAL_PARAMETER_GROUP_COMMON            (0 << 16)
+#define MMAL_PARAMETER_GROUP_COMMON		(0 << 16)
```
Here and a few other places include random whitespace changes. Those ought to be in a different patch.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h:
```diff
@@ -755,11 +811,11 @@ struct mmal_parameter_imagefx_parameters {
-	u32    max_height;
-	u32    lens_present;
-	u8     camera_name[MMAL_PARAMETER_CAMERA_INFO_MAX_STR_LEN];
+	u32 port_id;
```
Random whitespace

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h:
```diff
@@ -775,12 +831,106 @@ struct mmal_parameter_camera_info_flash_t {
 struct mmal_parameter_camera_info_t {
-	u32                            num_cameras;
-	u32                            num_flashes;
+	u32 num_cameras;
```
Random whitespace

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_fmts.h:
```diff
@@ -0,0 +1,278 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
```
Copyright & authorship missing.

naushir replied:
Done.

6by9 commented on include/uapi/linux/bcm2835-isp.h:
```diff
@@ -0,0 +1,317 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
```
Is that the right licence for userspace files? Doesn't it impose GPLv2 on the userspace app then?

videodev2.h is /* SPDX-License-Identifier: ((GPL-2.0+ WITH Linux-syscall-note) OR BSD-3-Clause) */

naushir replied:
Done.  I've followed videodev2.h here.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_ctrls.h:
```diff
@@ -0,0 +1,74 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Broadcom BM2835 ISP driver
```
BCM2835, not BM. (That's copied from the old bcm2835-camera driver which got it wrong in loads of places)

naushir replied:
Done

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+        :ref:`NV12/NV12 <V4L2-PIX-FMT-NV12>`,
+        :ref:`RGB565 <V4L2-PIX-FMT-RGB565>`,
+        :ref:`RGB24/BGR24 <V4L2-PIX-FMT-RGB24>`,
+        :ref:`ABGR32 <V4L2-PIX-FMT-ABGR32>`,
```
FYI I've just had a forum query as to why we do ABGR but not ARGB on the codecs (mainly encoder in). 
The main reason was that there wasn't a matching VC_IMAGE_T type, so it wasn't a quick mapping.

For encoder and ISP input it's input formatter config, so I'll look at adding it.
ISP output should be able to support other ordering too (I think it's only those two, not the other arrangements of the 4 components).

6by9 commented on Documentation/media/v4l-drivers/bcm2835-isp.rst:
```diff
@@ -0,0 +1,118 @@
+        :ref:`RAW12 <V4L2-PIX-FMT-SRGGB12>`,
+        :ref:`RAW12P <V4L2-PIX-FMT-SRGGB12P>`,
+        :ref:`RAW14 <V4L2-PIX-FMT-SRGGB14P>`,
+        :ref:`RAW16 <V4L2-PIX-FMT-SRGGB16>`.
```
And we support all the Bayer orders, so SRGGB16, SGRBG16, SGBRG16, and SBGGR16 for RAW16.

naushir replied:
The ref for the bayer formats only have one label, but in the actual documentation page it expands out all the orderings available.

6by9 reviewed (commented):
The review's coming piecemeal - all files done except the big one.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Broadcom BM2835 ISP driver
```
BCM2835

naushir replied:
Done

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+ *
+ * Copyright © 2019-2020 Raspberry Pi (Trading) Ltd.
+ *
+ * Author: Naushir Patuck @ Raspberry Pi
```
Would normally include an email address.
bcm2835-camera.c doesn't as most of those listed were dead due to people having moved on.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+#include <media/v4l2-ioctl.h>
+#include <media/videobuf2-vmalloc.h>
+#include <media/videobuf2-dma-sg.h>
+#include <media/videobuf2-dma-contig.h>
```
Why've we got vmalloc and dma-sg?
We're only using dma-contig allocation in this driver.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+/* Add one for the stats output node */
+#define BCM2835_ISP_NUM_NODES	\
+		(BCM2835_ISP_NUM_OUTPUTS + BCM2835_ISP_NUM_CAPTURES + 1)
```
Better to define it as ```NUM_METADATA_CAPTURES 1``` rather than a random +1

naushir replied:
Done

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+#define DEFAULT_DIM 1280U
+#define MIN_DIM 64U
+#define MAX_DIM 16384U
```
Comment stating what these defines are for? They're a touch terse otherwise.

Is 16k a max limit, and if so why?

naushir replied:
Done

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+#define INDEX_TO_NODE_TYPE(idx) ((idx) < BCM2835_ISP_NUM_NODES - 1 ?           \
+		 ((idx) < BCM2835_ISP_NUM_OUTPUTS ? NODE_TYPE_OUTPUT :         \
+		 NODE_TYPE_CAPTURE) : NODE_TYPE_STATS)
```
Don't each of the nodes have a V4L2_BUF_TYPE_? Do we need to create a new enum for this?

naushir replied:
True, I can reuse them and remote the custom type.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	struct vb2_queue queue;
+	unsigned int sequence;
+
+	/* The list of formats supported on input and output queues. */
```
"The list of formats supported on the node"?

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	struct bcm2835_isp_q_data q_data;
+	enum v4l2_colorspace colorspace;
+	unsigned int framerate_num;
+	unsigned int framerate_denom;
```
Do we need framerate per node? We don't allow rate adjustment, so this must be a global parameter, and presumably dictated by the OUTPUT queue (input to the device).

naushir replied:
You are right.  I'm inclined to remove framerate all together.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	bool queue_init;
+};
+
+#define node_get_dev(node) ((node)->dev)
```
inline function instead of a macro?

naushir replied:
Done.  I've converted NODE_IS_... to inline functions as well just to be consistent.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+	/* Image pipeline controls. */
+	int r_gain;
+	int b_gain;
```
Do we need our own local copies? Aren't they available by querying the ctrl_handler? ```v4l2_ctrl_find```?
(And I do acknowledge that bcm2835-camera keeps an internal copy)

naushir replied:
Yes, they are available in the ctrl handle via v4l2_ctrl_find.  However, it does seem easier to just hold onto the values right?

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+	return vchiq_mmal_port_parameter_set(dev->mmal_instance, port,
+					     parameter, value, value_size);
+}
```
Can't set_wb_gains and set_digital_gain use this instead of directly calling vchiq_mmal_port_parameter_set?

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	 * will reduce latency as the bistream parser will be kicked immediately
+	 * to parse the frame, rather than relying on its own heuristics for
+	 * when to wake up.
+	 */
```
Comment only applies to decoders. With raw video we always deal in full frames.

naushir replied:
Removed comment.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	 * Release the VCSM handle here as otherwise REQBUFS(0) aborts because
+	 * someone is using the dmabuf before giving the driver a chance to do
+	 * anything about it.
+	 */
```
Check whether this is still true. It may not be with V4L2_BUF_CAP_SUPPORTS_ORPHANED_BUFS now being supported by the core.

6by9 replied:
Checked, and it shouldn't abort now with orphaned buffers.
Amend the comment to something like
```
Release the VCSM handle here to release the associated dmabuf
```

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+static struct bcm2835_isp_fmt *get_default_format(struct bcm2835_isp_node *node)
+{
+	return &node->supported_fmts.list[node->id];
```
Why does the default format vary with the port number? That looks most odd but may be correct. I expected ```[0]```.

naushir replied:
Yes, that's a bug.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	.s_ctrl = bcm2835_isp_s_ctrl,
+};
+
+static struct v4l2_ctrl_config bcm2835_isp_custom_ctrl = {
```
Did you have to reuse bcm2835_isp_custom_ctrl as the variable name when you also have ```struct bcm2835_isp_custom_ctrl```?

It's also not used outside of ```register_node```, so why have it as a static struct instead of a local to that function? Seeing as it gets modified several times over, I'm assuming nothing takes a reference to it.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+/*
+ * Open one of the nodes /dev/video<N> associated with the ISP. Each node can be
+ * opened only once.
+ */
```
>Each node can be opened only once.

Where's that enforced?
And that means it'll fail v4l2-compliance as it checks that a node can be opened 100 times simultaneously.

Only one can stream at a time, but I'd expect all config to be common to all opened instances. You could be daft enough to open one instance to do all the control settings, and another to pass the buffers through.

naushir replied:
That comment is redundant, we do indeed allow multiple opens.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	v4l2_dbg(3, debug, &node_get_dev(node)->v4l2_dev,
+		 "Polling %s[%d]\n", node->name, node->id);
+
+	return vb2_poll(&node->queue, file, wait);
```
Do we really need this now for debug, or do we drop back to using vb2_fop_poll in the ops struct?

vb2_fop_poll  also uses a mutex, so are we thread safe here?

naushir replied:
No need for this any more, I will remove it.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	return vb2_mmap(&node->queue, vma);
```
Ditto. Use vb2_fop_mmap in the ops struct?

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	if (!NODE_IS_CAPTURE(node))
```
There would be a case to say that:
- g_fmt_vid_cap will have been disabled on OUTPUT and METADATA_CAPTURE nodes, 
- g_fmt_meta_cap will have been disabled on OUTPUT and CAPTURE nodes, 
- g_fmt_vid_out will have been disabled on CAPTURE and METADATA_CAPTURE nodes

Therefore there's no need to check the type here, and all instances could point use the same node lookup and ```populate_v4l_fmt``` call.

Or at least I expected them to be disabled via ```v4l2_disable_ioctl```, but I don't see such calls.

naushir replied:
Silly question I'm sure, but how do you disable g_fmt_vid_cap for output and metadata nodes using v4l2_disable_ioctl()? From what I can tell there is only one ioctl (VIDIOC_G_FMT) common to all node types that can be disabled, and we don't want to do that.

I could achieve similar by nulling out the relevant fn pointers in v4l2_ioctl_ops, for each node type.  That feels a bit clunky.

6by9 replied:
You're right.

Looking at v4l_enum_fmt in v4l2_ioctl.c it's switching on struct v4l2_fmtdesc type field and choosing the right handler function. The type field is passed in to the handler function too, so you can use the same function for capture, output, and meta_capture, and then check that
```
if f->type != node->type)
   return -EINVAL;
```

That works with enum_fmt. s_fmt appears to be the same too. I've not checked it through in g_fmt, and try_fmt, but I'd expect it to be true there as well.

naushir replied:
Thanks, I've consolidated all the node s_fmt, g_fmt, try_fmt, enum_fmt functions using the above test.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	}
+
+	if (!f->fmt.pix.colorspace || f->fmt.pix.colorspace == -1)
+		f->fmt.pix.colorspace = node->colorspace;
```
This is an odd clause. Can you explain the logic seeing as colorspace is actually an ```enum v4l2_colorspace```
0 would be V4L2_COLORSPACE_DEFAULT
-1 is invalid, particularly as f->fmt.pix.colorspace is a __u32. (Anything above 12 is invalid too).

Did we ever resolve whether we can sensibly select between BT601, BT709, and full/limited range on the output of the ISP?

naushir replied:
The -1 test is there for v4l2 compliance testing.  It sends (u32)-1 as a colourspace and expects the device driver to fix it.

No, the ISP does not currently handle different colourspaces.  It's easy to do for the colourspace matrix, but gamma is very tricky as we do not really follow any standard for gamma.  Rather we do what the AGC wants for  contrast control.

6by9 replied:
So it's expecting an override to some valid colorspace should it be invalid, which seems fair. But that means the check should be for it being greater than V4L2_COLORSPACE_DCI_P3 (which is the max permitted), or against the supported colorspace(s) for the pixel format requested on the port.

Better would be to do something sensible based on format (ie enforce V4L2_COLORSPACE_RAW for Bayer formats?) and V4L2_COLORSPACE_SMPTE170M/V4L2_COLORSPACE_REC709 for YUV (switch based on res > 720x576 typically, or hard code for now based on the CCM/gamma), and V4L2_COLORSPACE_SRGB for RGB.

Comments in the header do say
```
	/*
	 * Default colorspace, i.e. let the driver figure it out.
	 * Can only be used with video capture.
	 */
	V4L2_COLORSPACE_DEFAULT       = 0,
```
so it should only be applicable against the output side of the ISP.

naushir replied:
So I think I'll do is assign a colourspace to each node format (RGB->V4L2_COLORSPACE_SRGB, Bayer->V4L2_COLORSPACE_RAW and YUV->V4L2_COLORSPACE_SMPTE170M).

I will update any user requested colourspace with one of the three above based on format.

I do not think we should make the switch between V4L2_COLORSPACE_SMPTE170M and V4L2_COLORSPACE_REC709 as the ISP params do not actually change with the colourspace currently, it sets SMPTE170M/BT601 unconditionally.

6by9 replied:
Sounds good.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+		/* Format found */
+		fmt = &fmts->list[f->index];
+		f->pixelformat = fmt->fourcc;
+		f->flags = fmt->flags;
```
Do we really need to copy into fmt? The lines don't look that long, so
```
		f->pixelformat = fmts->list[f->index]->fourcc;
		f->flags = fmts->list[f->index]->flags;
```
and drop ```fmt```.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	if (!NODE_IS_CAPTURE(node))
```
Same discussion as for g_fmt variants. If you've disabled them then there's no need for unique functions.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+		return -EINVAL;
+
+	return vidioc_try_fmt(node, f, fmt);
+}
```
Same discussion as for g_fmt variants. If you've disabled them then there's no need for unique functions.

try_fmt_meta_cap is different, but I don't see why seeing as the default (only) format should be V4L2_META_FMT_BCM2835_ISP_STATS. Perhaps it's because ```get_default_format``` doing odd things as noted above.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
```
And once more. Disable the invalid ioctls per node, and then you can use one common function for all 3 s_fmt functions, calling the common try_fmt.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+
+	/* This return value is required fro V4L2 compliance. */
+	if (NODE_IS_STATS(node))
+		return -ENOTTY;
```
In which case the ioctl should have been disabled.

naushir replied:
Sadly it seems that even if I disable the ioctl for the stats node, v4l2-compliance complains as the framework does not return -ENOTTY.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	if (NODE_IS_STATS(node))
+		return -ENOTTY;
+
+	if (!NODE_IS_OUTPUT(node) || s->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
```
The first case should be done by disabling the ioctl.
The second case needs to remain.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+	switch (s->type) {
+	/* Cannot change source parameters dynamically at runtime. */
+	case V4L2_EVENT_SOURCE_CHANGE:
+		return -EINVAL;
```
Do we need to reject it then?
Let it pass into the default clause and the framework will handle it, but nothing will ever generate the event.

naushir replied:
It seems you do need to explicitly reject it.  Otherwise the compliance test fails if I let it pass through he default clause.

From what I can tell, the compliance test is expecting this specific behaviour for m2m devices.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+module_platform_driver(bcm2835_isp_pdrv);
+
+MODULE_DESCRIPTION("BCM2835 ISP driver");
+MODULE_AUTHOR("Naushir Patuck");
```
Email address.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1808 @@
+		node->id = index;
+		node->vfl_dir = VFL_DIR_TX;
+		node->name = "output";
+		break;
```
v4l2_disable_ioctl calls in these cases to disable all the invalid S_FMT, G_FMT, TRY_FMT, ENUM_FMT etc calls.

6by9 replied:
Ack that we can't disable the individual S_FMT etc types, but that's been resolved.

What about S_SELECTION? (I think all the others fom bcm2835_isp_node_ioctl_ops are valid)

naushir replied:
This is done lower in the function.

(as per the comment lower, it will be moved up to the switch statement).

6by9 reviewed (commented):
First pass on the main driver. Nothing else immediately stands out, but there are a fair number of reworks suggested in there.

naushir committed:
staging: mmal-vchiq: Fix formatting errors in mmal_parameters.h

No functional changes in this commit.

- Remove erroneous whitespace.
- Remove _t postfix label on structs and enums.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
staging: vc04_services: ISP: Add a more complex ISP processing component

Driver for the BCM2835 ISP hardware block.  This driver uses the MMAL
component to program the ISP hardware through the VC firmware.

The ISP component can produce two video stream outputs, and Bayer
image statistics. This can't be encompassed in a simple V4L2
M2M device, so create a new device that registers 4 video nodes.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
uapi: bcm2835-isp: Add bcm2835-isp uapi header file

This file defines the userland interface to the bcm2835-isp driver
that will follow in a separate commit.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir committed:
media: uapi: v4l2-core: Add ISP statistics output V4L2 fourcc type

Add V4L2_META_FMT_BCM2835_ISP_STATS V4L2 format type.

This new format will be used by the BCM2835 ISP device to return
out ISP statistics for 3A.

Signed-off-by: Naushir Patuck <naush@raspberrypi.com>

naushir reviewed (commented):
Think I've covered the list.  Updates incoming.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/Kconfig:
```diff
@@ -0,0 +1,14 @@
+	depends on VIDEO_V4L2 && (ARCH_BCM2835 || COMPILE_TEST)
+	select BCM2835_VCHIQ_MMAL
+	select VIDEOBUF2_DMA_CONTIG
+	select V4L2_MEM2MEM_DEV
```
Do we actually use anything from V4L2_MEM2MEM_DEV, or is this a leftover from bcm2835-codec?

naushir replied:
No, not needed any more.

6by9 replied:
Actually don't we have a dependency on MEDIA_CONTROLLER here too?
I've lost track of which things are automatically selected and which aren't these days. Perhaps enough of MEDIA_CONTROLLER is stubbed when it isn't acutally built such that it works.

naushir replied:
I assumed MEDIA_SUPPORT pulled in the bits of media controller.

Seems to build correctly regardless.

6by9 replied:
MEDIA_SUPPORT != MEDIA_CONTROLLER, it's any part of V4L2/DVB/Radio/SDR/etc.
It's defined in drivers/media/Kconfig, ie the base directory of all media.

It does appear that all the function prototypes in media-device.h are stubbed should CONFIG_MEDIA_CONTROLLER not be defined.
https://elixir.bootlin.com/linux/latest/source/include/media/media-device.h#L440

naushir replied:
Added MEDIA_CONTROLLER.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_fmts.h:
```diff
@@ -0,0 +1,298 @@
+#define BCM2835_ISP_FMTS
+
+#include "vchiq-mmal/mmal-encodings.h"
+
```
You've only included the MMAL header, so if this is included before the header defining V4L2_PIX_FMT_* (videodev2.h) then you'll have a compile failure.

naushir replied:
Done

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_fmts.h:
```diff
@@ -0,0 +1,298 @@
+	u32 flags;
+	u32 mmal_fmt;
+	int size_multiplier_x2;
+	int colorspace;
```
int? enum v4l2_colorspace surely.

naushir replied:
Done

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_fmts.h:
```diff
@@ -0,0 +1,298 @@
+		.size_multiplier_x2 = 3,
+		.colorspace	    = V4L2_COLORSPACE_SMPTE170M,
+	},
+	{
```
IIRC linux-media coding style prefers
```
}, {
```
on one line.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h:
```diff
@@ -780,7 +836,101 @@ struct mmal_parameter_camera_info_t {
+		flashes[MMAL_PARAMETER_CAMERA_INFO_MAX_FLASHES];
+};
+
+struct mmal_parameter_ccm_t {
```
I know I'm guilty above, but kernel style normally doesn't include ```_t``` suffixes. They're not typdef'ed, therefore you have the ```struct``` telling you the type anyway.

naushir replied:
Yeah, I wanted to keep the format consistent with the rest of the file.  Should I take the hit and just fixup all the _t usage, or just fixup my additions?

6by9 replied:
Most of the file is without the _t. It's only the last couple that have it.
Fix them up as a separate patch, then add your new controls.

naushir replied:
I'll add the_t fixes to the same patch that fixes whitespace in the file before adding the new controls then.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_ctrls.h:
```diff
@@ -0,0 +1,74 @@
+		.flags  = 0
+	},
+	{
+		.name	= "GEQ",
```
Ditto
I'd love an ISP with a Graphic Equaliser.....

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_ctrls.h:
```diff
@@ -0,0 +1,74 @@
+
+static const struct bcm2835_isp_custom_ctrl custom_ctrls[] = {
+	{
+		.name	= "CCM",
```
Very terse name.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835_isp_ctrls.h:
```diff
@@ -0,0 +1,74 @@
+		.flags  = 0
+	},
+	{
+		.name	= "DPC",
```
Ditto

naushir replied:
Done.

6by9 reviewed (commented):
Two parts again - this is all files except bcm2835-v4l2-isp.c

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1669 @@
+		if (fmt->fourcc == (node_is_stats(node) ?
+					    f->fmt.meta.dataformat :
+					    f->fmt.pix.pixelformat))
+			break;
```
(I know this is my code from bcm2835-codec, but I'm now reviewing it as yours!)
Can't we return ```return fmt;``` here, which makes the end of the function just ```return NULL;```?

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1669 @@
+	switch (ctrl->id) {
+	case V4L2_CID_RED_BALANCE:
+		dev->r_gain = ctrl->val;
+		if (set_wb_gains(node))
```
All of these setters don't set up ret based on the return code, so rely on the initialiser above.
Is there gain in replacing with
```
int ret;
switch(ctrl->id)
{
case V4L2_CID_RED_BALANCE:
   dev->r_gain = ctrl->val;
   ret = set_wb_gain(node);
   break;
...
default:
   ret = -EINVAL;
}
if (ret)
			v4l2_err(&dev->v4l2_dev, "%s: Failed setting ctrl  \"%s\" (%08x), err %d\n",
				 __func__, ctrl->name, ctrl->id, ret);
```

naushir replied:
Done.  Does make the code look a bit neater.

naushir replied:
Had to make a slight change here:

```
if (ret) {
   v4l2_err(&dev->v4l2_dev, "%s: Failed setting ctrl  \"%s\" (%08x), err %d\n",
				 __func__, ctrl->name, ctrl->id, ret);
   ret = -EIO
}
```

Otherwise v4l2-compliance complains that the mmal return error code is invalid.  -EIO is acceptable but does flag as a warning.

The reason for the mmal failure is the LS parameters are invalid (all zeros) and it correctly rejects it.  Given this is a compound control and cannot have any min/max/default, I feel v4l2-compliance should not expect arbitrary data filled into the control to actually pass...

6by9 replied:
Yup, one of the cleanups to do at some point is to get rid of all the VCHIQ error numbers, and map MMAL errors to kernel ones. Adding ```ret = -EIO``` is fine. -EINVAL is probably not allowed as the ctrl handler is supposed to have sanity checked most of the parameters anyway.

I'm assuming you're only seeing a compliance failure rather than lockup. Yes, it seems to be an oversight in compliance if it tries testing compound controls that can't have been sanity checked.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1669 @@
+		f->fmt.pix.height = max(min(f->fmt.pix.height, MAX_DIM),
+					MIN_DIM);
+
+		fmt = find_format(f, node);
+		if (!fmt) {
+			f->fmt.pix.pixelformat =
+					get_default_format(node)->fourcc;
+			fmt = find_format(f, node);
+		}
```
This is common to both paths (get_default_format should return V4L2_META_FMT_BCM2835_ISP_STATS on the stats node), so could be before the if.
Do the other fields need zero'ing on the stats node? For compressed formats width/height etc are 0, with only sizeimage being set. STATS is effectively equivalent to compressed data. (Almost query whether it should have the compressed flag set).

naushir replied:
If I understand your comment correctly, it would be a bit awkward to do this.   f->fmt is a union, and the .pix field is used for images and .meta is used for metadata.  If I were to combine the two paths, I will still end up with multiple if clauses (to setup the .pix.width and .pix.height before the find_format).

For the .meta field in the stats node (given it's a union) can only touch the .dataformat and .buffersize fields.

6by9 replied:
Ah, I'd missed the union. 
You can still make it
```
fmt = find_format(f, node);
if (!fmt)
  fmt = get_default_format(node);
if (!node_is_stats(node)) {
  f->fmt.pix.pixelformat = fmt->fourcc;
 ...
} else {
 f->fmt.meta.dataformat = fmt->fourcc;
 ...
}
```
If the lookup succeeded then fmt is the correct format, otherwise you've chosen the default.
```

naushir replied:
The code would have to look something like:

```
if (!node_is_stats(node)) {
		f->fmt.pix.width = max(min(f->fmt.pix.width, MAX_DIM), MIN_DIM);
		f->fmt.pix.height = max(min(f->fmt.pix.height, MAX_DIM), MIN_DIM);
}

fmt = find_format(f, node);
if (!fmt)
  fmt = get_default_format(node);
if (!node_is_stats(node)) {
  f->fmt.pix.pixelformat = fmt->fourcc;
 ...
} else {
 f->fmt.meta.dataformat = fmt->fourcc;
 ...
}`
```
So we still have the extra if (!stats) block at the start.  However, I could simplify the original code by doing somthing like:

```
if (!stats) {
...
} else {
   f->fmt.meta.dataformat = V4L2_META_FMT_BCM2835_ISP_STATS;
   f->fmt.meta.buffersize = get_port_data(node)->minimum_buffer.size;
}

```

i.e. the stats clause does not need to go through find_format as we konw there is only going to be one stats format.

I'll use your judgement on what you think might be the better way.

6by9 replied:
I'd go for 
```
	fmt = find_format(f, node);
	if (!fmt)
		fmt = get_default_format(node);

	if (!node_is_stats(node)) {
		f->fmt.pix.width = max(min(f->fmt.pix.width, MAX_DIM),
				       MIN_DIM);
		f->fmt.pix.height = max(min(f->fmt.pix.height, MAX_DIM),
					MIN_DIM);

		f->fmt.pix.pixelformat = fmt->fourcc;

		f->fmt.pix.colorspace = fmt->colorspace;
		f->fmt.pix.bytesperline = get_bytesperline(f->fmt.pix.width,
							   fmt);
		f->fmt.pix.field = V4L2_FIELD_NONE;
		f->fmt.pix.sizeimage =
			get_sizeimage(f->fmt.pix.bytesperline, f->fmt.pix.width,
				      f->fmt.pix.height, fmt);
	} else {
		f->fmt.meta.dataformat = fmt->fourcc;
		f->fmt.meta.buffersize =
				get_port_data(node)->minimum_buffer.size;
	}
```

naushir replied:
Done

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1669 @@
+		return -ENOTTY;
+
+	if (s->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
+		return -EINVAL;
```
This feels odd after a check of the node type.

Does this mean that setting a type != V4L2_BUF_TYPE_VIDEO_OUTPUT on either OUTPUT or CAPTURE nodes gets accepted? I'd expect only setting with type = V4L2_BUF_TYPE_VIDEO_OUTPUT && on the OUTPUT node would be required.

naushir replied:
This was down to the compliance failing.  However, with the relevant ioctls disabled, it's not needed any more.

naushir replied:
Actually, it's still a problem with compliance that forces an invalid value into s->type.  I have to keep the clause in there.

6by9 replied:
It felt odd rather than being definitely wrong. I'd half suggest ```if (s->type != node->type)``` again, but I don't see it ever getting enabled on the CAPTURE queues.

naushir replied:
Sigh, I think I understand what happened.  Summary is that I need to disable the old crop API ioctls explicitly, otherwise v4l2-compliance (and the v4l2 framework) does some strange things and goes into the s/g_selection function **even though they have been disabled**.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1669 @@
+	.vidioc_try_fmt_vid_cap		= bcm2835_isp_node_try_fmt,
+	.vidioc_try_fmt_vid_out		= bcm2835_isp_node_try_fmt,
+	.vidioc_try_fmt_meta_cap	= bcm2835_isp_node_try_fmt,
+	.vidioc_s_selection		= bcm2835_isp_node_s_selection,
```
What about g_selection?

naushir replied:
Added.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1669 @@
+	if (!node_is_output(node))
+		v4l2_disable_ioctl(&node->vfd, VIDIOC_S_CTRL);
+	if (node_is_capture(node))
+		v4l2_disable_ioctl(&node->vfd, VIDIOC_S_SELECTION);
```
Too late.
https://elixir.bootlin.com/linux/latest/source/include/media/v4l2-dev.h#L459
```
 *    This must be called before video_register_device.
 *    See also the comments for determine_valid_ioctls().
```
You called video_register_device at line 1312.
Put it in the switch/case above.

naushir replied:
Done.

6by9 reviewed (commented):
A load more comments.

I just tried v4l2-compliance against an old firmware (Pi3 Feb 2020), and it crashed the task such that I couldn't ctrl-C it. It was whilst doing the ctrl ioctl tests, and "set WB gains" had failed.

At a guess it may also be an old enough ISP component to have fewer output ports. A check of the port configuration after vchiq_mmal_component_init in bcm2835_isp_probe wouldn't go amiss.

6by9 commented:
Almost guaranteed to be only 2 output ports. Splat during probe:
```
[    8.720847] 8<--- cut here ---
[    8.726760] Unable to handle kernel NULL pointer dereference at virtual address 00000004
[    8.726773] pgd = (ptrval)
[    8.726783] [00000004] *pgd=00000000
[    8.726803] Internal error: Oops: 5 [#1] SMP ARM
[    8.726816] Modules linked in: vc4(+) brcmutil drm_kms_helper drm drm_panel_orientation_quirks sha256_generic libsha256 i2c_mux_pinctrl snd_soc_core i2c_mux snd_compress cfg80211 snd_pcm_dmaengine syscopyarea sysfillrect raspberrypi_hwmon sysimgblt fb_sys_fops hwmon rfkill bcm2835_codec(C+) bcm2835_isp(C+) bcm2835_v4l2(C) v4l2_mem2mem bcm2835_mmal_vchiq(C) videobuf2_dma_contig videobuf2_vmalloc videobuf2_memops videobuf2_v4l2 videobuf2_common snd_bcm2835(C) snd_pcm i2c_bcm2835 snd_timer videodev snd mc vc_sm_cma(C) uio_pdrv_genirq uio fixed i2c_dev ip_tables x_tables ipv6 nf_defrag_ipv6
[    8.742547] brcmfmac: F1 signature read @0x18000000=0x15264345
[    8.745636] CPU: 1 PID: 154 Comm: systemd-udevd Tainted: G         C        5.4.32-v7+ #4
[    8.745642] Hardware name: BCM2835
[    8.745674] PC is at port_parameter_get+0x40/0x114 [bcm2835_mmal_vchiq]
[    8.745685] LR is at 0x81ebce00
[    8.776078] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43455-sdio for chip BCM4345/6
[    8.782820] bcm2835-codec bcm2835-codec: Device registered as /dev/video10
[    8.782867] bcm2835-codec bcm2835-codec: Loaded V4L2 decode
[    8.832946] pc : [<7f2b2694>]    lr : [<81ebce00>]    psr: 60000113
[    8.832953] sp : b933b7a0  ip : 00000000  fp : b933b9e4
[    8.832958] r10: 00000023  r9 : b933ba34  r8 : b933b7b0
[    8.832965] r7 : b933ba30  r6 : 00000000  r5 : 80d04f48  r4 : b7140000
[    8.832971] r3 : b933ba34  r2 : 00000001  r1 : 0000000f  r0 : b7140000
[    8.832984] Flags: nZCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment user
[    8.832991] Control: 10c5383d  Table: 3926406a  DAC: 00000055
[    8.833000] Process systemd-udevd (pid: 154, stack limit = 0x(ptrval))
[    8.833007] Stack: (0xb933b7a0 to 0xb933c000)
[    8.833020] b7a0: b933b7dc b933b7b0 80157910 8015453c 00000000 b73485a0 b93f0000 0000000f
[    8.833032] b7c0: b93f1e00 b933b84c 80da3960 80d05684 b933b7f4 b933b7e0 8014d640 80157824
[    8.833044] b7e0: b93f1e00 babcd700 b933b824 b933b7f8 8014d670 8014d5b8 00000000 00000000
[    8.833062] b800: b93f1e00 babcd700 00000000 b933b84c 60000193 b93ca0fc b933b844 b933b828
[    8.843738] usbcore: registered new interface driver brcmfmac
[    8.857209] b820: 8010cfcc 8010bb98 8010bbbc 80d04f48 00000000 00000001 b933b87c b933b848
[    8.857221] b840: 8010bc18 8010cfb4 8010bc18 00000000 b933b87c b933b860 8010cfcc 8010bb98
[    8.857235] b860: 8010bbbc 80d04f48 00000000 00000000 b933b8b4 b933b880 8010bc18 8010cfb4
[    8.857247] b880: 8010bc18 00000000 8016a8f4 80c9bb10 808c4a7c 00000000 8016a8f4 b933b908
[    8.857259] b8a0: 00000000 60000113 b933b8dc b933b8b8 8020b004 8020bd80 00000393 60000113
[    8.857270] b8c0: 00000001 00000000 00000003 b933b908 b933b8f4 b933b8e0 808c4a7c 8020afb8
[    8.857282] b8e0: 80d8b97c 80d8b978 b933b94c b933b8f8 80dcf176 80dcf50c b933b944 7f2f01b3
[    8.857294] b900: b933b98c b933b910 808b7688 808b5680 ffffff04 7f2f01b7 80dcf50c 00000000
[    8.940321] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac43455-sdio.raspberrypi,3-model-b-plus.txt failed with error -2
[    8.944931] b920: 80d04f48 ffffffff 00000000 00000000 00000000 ffffffff 0000000a 00000001
[    8.944944] b940: 7f2f01b8 80dc3135 b933b984 b933b960 808b64f8 808b6238 b933b9bc 81ebce00
[    9.260892] b960: b933b98c b933b970 8010cfcc 8010bb98 8010bbbc 80d04f48 00000000 00000119
[    9.278131] b980: b933b9c4 b933b990 8010bc18 8010cfb4 b933b9b4 00000000 8017f15c 80c9bb10
[    9.295430] b9a0: 8017d0f0 808bf548 00000080 b7140004 b933b9cc b933b9c0 808bf548 801901d4
[    9.312727] b9c0: b7140000 b7140004 b933ba34 00000001 b7140560 00000003 b933ba14 b933b9e8
[    9.330023] b9e0: 7f2b27c0 7f2b2660 b933ba30 8020afb8 8020b004 b7140000 80d04f48 ba6af000
[    9.347321] ba00: b93ba040 b93bb540 b933bb04 b933ba18 7f2edad4 7f2b2774 b933ba30 b933ba28
[    9.364622] ba20: 80229294 80d04f48 b93bb1e0 b93ba040 000000a0 b933ba40 8017ef3c 8022927c
[    9.381922] ba40: b933ba94 b933ba50 8017f160 8017eed8 7f2f0180 b933bafc 7f2ef4c8 b93badf8
[    9.399222] ba60: 0000000f 00000000 b933ba84 80c96408 b933bafc 7f2f0180 b93ba040 b93bb010
[    9.416522] ba80: 00000002 00000001 b933baac b933ba98 8017f32c 8017ef4c 7f2f0180 b933bafc
[    9.433804] baa0: b933bad4 b933bab0 801800e4 8017f2e8 7f2f1080 00000000 80d04f48 80d04f48
[    9.451073] bac0: 00000000 b93ba040 b933baf4 b933bad8 8017f928 81ebce00 b933bafc b93bb1e0
[    9.468339] bae0: 80d04f48 ba6af000 b93ba040 b93bb540 00000003 00000023 b933bbd4 b933bb08
[    9.485605] bb00: 7f2ede4c 7f2eda78 0000000f 00000001 0000ffff 00000000 00000001 00000000
[    9.502879] bb20: 000003e8 00000000 00000100 b93bb040 b93ba09c 7f2ef760 7f2ef6d4 00000000
[    9.520141] bb40: 009819c8 7f2f0384 00000100 00000000 00000000 00000000 000000ff 00000000
[    9.537397] bb60: 00000001 00000000 00000000 00000000 00000008 00000000 00000000 00000000
[    9.554659] bb80: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    9.571919] bba0: b7140004 81ebce00 7f12a460 b93ba040 00000000 ba6af000 b93bb6c0 00000003
[    9.589180] bbc0: 7f2f1018 00000023 b933bc24 b933bbd8 7f2ee2e0 7f2edda8 80b0964c 00000001
[    9.606451] bbe0: 7f2f1018 00000023 b933bc1c b933bbf8 8037a698 81ebce00 00000001 00000000
[    9.623711] bc00: ba6af010 7f2f1018 80e4bbc4 00000000 7f2f1018 00000023 b933bc44 b933bc28
[    9.640969] bc20: 80609940 7f2ee244 806098c4 80e4bbc0 ba6af010 00000000 b933bc74 b933bc48
[    9.658227] bc40: 80607988 806098d0 00000000 ba6af010 7f2f1018 7f2f1018 80d04f48 00000000
[    9.675488] bc60: b9eaeac0 b9eaeae4 b933bcb4 b933bc78 80607cf8 806077a8 b933bc9c b933bc88
[    9.692753] bc80: 808c1e1c 808bf524 00000000 ba6af010 00000000 ba6af010 7f2f1018 80d04f48
[    9.710012] bca0: 00000000 b9eaeac0 b933bcd4 b933bcb8 8060802c 80607c94 8060a1a0 00000000
[    9.727274] bcc0: 7f2f1018 ba6af010 b933bcf4 b933bcd8 806080c4 80607fd0 00000816 00000000
[    9.744536] bce0: 7f2f1018 80608034 b933bd24 b933bcf8 80605708 80608040 b933bd30 ba4c0e58
[    9.761797] bd00: b9e36534 81ebce00 7f2f1018 80d6a868 b8259f00 00000000 b933bd34 b933bd28
[    9.779083] bd20: 806071e4 8060569c b933bd5c b933bd38 80606b6c 806071c4 7f2f0030 b933bd48
[    9.796341] bd40: 7f2f1018 7f0a1000 b933a000 80d04f48 b933bd74 b933bd60 8060877c 80606a38
[    9.813602] bd60: 7f2f1080 7f0a1000 b933bd84 b933bd78 80609870 806086fc b933bd94 b933bd88
[    9.830874] bd80: 7f0a1024 8060982c b933be0c b933bd98 801031c8 7f0a100c b933bdd4 808bf548
[    9.848133] bda0: 00000080 ba401e40 b933bdc4 b933bdb8 808bf548 802d4b04 b933be0c b933bdc8
[    9.865391] bdc0: 802d4b04 8029d33c b933be04 b933bdd8 802b8d68 00001ac7 b9eaebc0 a0000113
[    9.882650] bde0: 808aacf8 81ebce00 7f2f1080 b9eaeb80 80d04f48 00000001 00000001 b9eaeac0
[    9.899911] be00: b933be34 b933be10 801be798 80103184 b933be34 b933be20 802b8edc b933bf28
[    9.917174] be20: 7f2f1080 80d04f48 b933bf04 b933be38 801bd6c8 801be734 7f2f108c 00007fff
[    9.934445] be40: 7f2f1080 801ba364 00000000 7f0a43d1 7f2f1274 7f2f1194 80d04f48 7f0a2000
[    9.951705] be60: b9eaeac8 7f2f10c8 b933be8c 80b478b0 802e0a50 802e08b0 b933be98 80d04f48
[    9.968966] be80: 00000000 00000000 00000000 00000000 00000000 00000000 65646976 6675626f
[    9.986222] bea0: 34765f32 0000326c 00000000 00000000 00000000 00000000 00000000 00000000
[   10.003483] bec0: 00000000 00000000 00000000 00000000 00000000 81ebce00 7fffffff 80d04f48
[   10.020750] bee0: 00000000 00000015 76d568e0 7fffffff 00000000 0000017b b933bfa4 b933bf08
[   10.038022] bf00: 801bde70 801bb85c 7fffffff 00000000 00000003 00000000 76d568e0 bbb7b000
[   10.055306] bf20: 00009378 00000000 bbb7e2d1 bbb80600 bbb7b000 00009378 bbb83e50 bbb808eb
[   10.072600] bf40: bbb81dfc 00006000 00006bc0 000032ec 000072e3 00000000 00000000 00000000
[   10.089902] bf60: 000032dc 0000001f 00000020 00000016 00000012 0000000c 00000000 81ebce00
[   10.107211] bf80: a2cff600 00000000 00000000 0000017b 801011c4 b933a000 00000000 b933bfa8
[   10.124534] bfa0: 801011a4 801bddc0 a2cff600 00000000 00000015 76d568e0 00000000 76d573f4
[   10.141861] bfc0: a2cff600 00000000 00000000 0000017b 0143fc78 004d11dc 01437b10 00000000
[   10.159204] bfe0: 7e829180 7e829170 76d4d9d8 76e3daf0 60000010 00000015 00000000 00000000
[   10.176543] Backtrace: 
[   10.183425] [<7f2b2654>] (port_parameter_get [bcm2835_mmal_vchiq]) from [<7f2b27c0>] (vchiq_mmal_port_parameter_get+0x58/0x78 [bcm2835_mmal_vchiq])
[   10.205761]  r9:00000003 r8:b7140560 r7:00000001 r6:b933ba34 r5:b7140004 r4:b7140000
[   10.218157] [<7f2b2768>] (vchiq_mmal_port_parameter_get [bcm2835_mmal_vchiq]) from [<7f2edad4>] (bcm2835_isp_get_supported_fmts+0x68/0x330 [bcm2835_isp])
[   10.241037]  r8:b93bb540 r7:b93ba040 r6:ba6af000 r5:80d04f48 r4:b7140000
[   10.252307] [<7f2eda6c>] (bcm2835_isp_get_supported_fmts [bcm2835_isp]) from [<7f2ede4c>] (register_node+0xb0/0x49c [bcm2835_isp])
[   10.273044]  r10:00000023 r9:00000003 r8:b93bb540 r7:b93ba040 r6:ba6af000 r5:80d04f48
[   10.285447]  r4:b93bb1e0
[   10.292369] [<7f2edd9c>] (register_node [bcm2835_isp]) from [<7f2ee2e0>] (bcm2835_isp_probe+0xa8/0x52c [bcm2835_isp])
[   10.311880]  r10:00000023 r9:7f2f1018 r8:00000003 r7:b93bb6c0 r6:ba6af000 r5:00000000
[   10.324259]  r4:b93ba040
[   10.331147] [<7f2ee238>] (bcm2835_isp_probe [bcm2835_isp]) from [<80609940>] (platform_drv_probe+0x7c/0xb4)
[   10.349703]  r10:00000023 r9:7f2f1018 r8:00000000 r7:80e4bbc4 r6:7f2f1018 r5:ba6af010
[   10.362068]  r4:00000000
```

naushir commented:
> At a guess it may also be an old enough ISP component to have fewer output ports. A check of the port configuration after vchiq_mmal_component_init in bcm2835_isp_probe wouldn't go amiss.

Latest revision (incoming) adds a test to ensure we have the correct number of i/p and o/p ports.

naushir commented:
Update to remove erroneous comment and replace macro with an inline function.

naushir commented:
> Mayby I'm fault, but this commit breaks building kernel fpr Raspberry Pi1
> 
> Commenting out `obj-$(CONFIG_CLK_BCM2835) += clk-bcm2711-dvp.o` fixes it termporarly

That's an interesting observation - I will look into it.
This PR does not touch that file, so I can't tell straight away what could be wrong.  Could you post the compile error message please?

6by9 commented:
>> Mayby I'm fault, but this commit breaks building kernel fpr Raspberry Pi1
>>
>>  Commenting out obj-$(CONFIG_CLK_BCM2835) += clk-bcm2711-dvp.o fixes it termporarly

>That's an interesting observation - I will look into it.

Surely that's a comment against the wrong PR as this doesn't touch it. I've kicked off a build to see what errors it produces.
@naushir Please don't get sidetracked onto this. I'll review this again now, but clk-bcm2711-dvp is not a blocker for merging it.

6by9 commented:
The comment is against a commit too, not against this PR. I think this is a Github glitch.

naushir commented:
Confirmed my build is fine.  Won't be looking into this any more, as @6by9 mentioned, this is probably for a different PR.

6by9 commented on include/uapi/linux/bcm2835-isp.h:
```diff
@@ -0,0 +1,319 @@
+struct bcm2835_isp_gamma {
+	__u32 enabled;
+	__u16 X[BCM2835_NUM_GAMMA_PTS];
+	__u16 Y[BCM2835_NUM_GAMMA_PTS];
```
Kernel coding style wouldn't normally use capitals.

naushir replied:
Done.

6by9 commented on MAINTAINERS:
```diff
@@ -3212,6 +3212,12 @@ S:	Maintained
+L:	linux-media@vger.kernel.org
+S:	Maintained
+F:	drivers/staging/vc04_services/bcm2835-isp
+
```
This ought to include the uapi header as well.

naushir replied:
Done.  Also looks like I should be adding the documentation.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	gains.b_gain.num = dev->b_gain;
+	gains.r_gain.den = 1000;
+	gains.b_gain.den = 1000;
+
```
(nit) Swap the two variable defns around, and you can set this all up with initialisers along the lines of:
```
	struct bcm2835_isp_dev *dev = node_get_dev(node);
	struct mmal_parameter_awbgains gains = { 
		.r_gain = { dev->r_gain, 1000 },
		.b_gain = { dev->b_gain, 1000 }
	};
	return set_isp_param....
```

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	struct mmal_parameter_rational digital_gain;
+
+	digital_gain.num = gain;
+	digital_gain.den = 1000;
```
Ditto can be done with initialisers.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	port->cb_ctx = node;
+	ret = vchiq_mmal_port_enable(dev->mmal_instance, port,
+				     mmal_buffer_cb);
+	if (ret == 0)
```
```!ret```

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	if (ret) {
+		v4l2_err(&dev->v4l2_dev, "v4l2_fh_open failed\n");
+		goto unlock_return;
+	}
```
Why are we having to claim a lock here?
I only see it being claimed and released in _open and _release, and v4l2_fh_open appears to be thread safe anyway (it's not accessing any of our node info).

Can we not just use v4l2_fh_open and vb2_fop_release as the open and release functions in bcm2835_isp_fops?

naushir replied:
Yes, no need for the debug log points.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+		 "%s: Calculated bpl as %u, size %u\n", __func__,
+		 q_data->bytesperline, q_data->sizeimage);
+
+	/* If we have a component then setup the port as well */
+	port = get_port_data(node);
+	if (!port)
+		return 0;
```
I think this is left over from bcm2835-codec where we create a component on start_streaming. You create your 1 component in probe, so port will always be non-NULL.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_dev *dev = node_get_dev(node);
+	struct vchiq_mmal_port *port = get_port_data(node);
+	u32 crop_size = sizeof(crop);
```
(nit) Longest to shortest, so move crop to here.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	u32 crop_size = sizeof(crop);
+	int ret;
+
+	/* This return value is required fro V4L2 compliance. */
```
s/fro/for

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+		 "Stream on for node %s[%d]\n", node->name, node->id);
+
+	return vb2_streamon(&node->queue, type);
+}
```
Hardly worth having the function for the debug message

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+
+	vb2_streamoff(&node->queue, type);
+	return 0;
+}
```
Hardly worth having the function for the debug message

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+		num_encodings = param_size / sizeof(u32);
+	}
+
+	/* Assume at this stage that all encodings will be supported in V4L2.
```
(nit) /* on separate line to first comment line.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+{
+	struct video_device *vfd;
+	struct vb2_queue *queue;
+	int i, ret;
```
unsigned int i

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+
+static void media_controller_unregister(struct bcm2835_isp_dev *dev)
+{
+	int i;
```
unsigned int i

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+static int media_controller_register(struct bcm2835_isp_dev *dev)
+{
+	char *name;
+	int ret, i = 0;
```
unsigned int i

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+static int bcm2835_isp_remove(struct platform_device *pdev)
+{
+	struct bcm2835_isp_dev *dev = platform_get_drvdata(pdev);
+	int i;
```
unsigned int i

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+static int bcm2835_isp_probe(struct platform_device *pdev)
+{
+	struct bcm2835_isp_dev *dev;
+	int ret, i = 0;
```
unsigned int i

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	.remove = bcm2835_isp_remove,
+	.driver = {
+			.name = BCM2835_ISP_NAME,
+			.owner = THIS_MODULE,
```
owner not needed.

naushir replied:
Removed.

6by9 commented on drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h:
```diff
@@ -780,7 +836,101 @@ struct mmal_parameter_camera_info {
+	u16 black_level_r;
+	u16 black_level_g;
+	u16 black_level_b;
+};
```
Is it really defined as being 10 bytes total, so not a multiple of 4 ? Oh well, but a padding field would have been nice to avoid oddballs.
The firmware will be checking whether the struct is large enough, so there is the potential for rejection if the compilers differ in padding behaviour.

naushir replied:
I've added a padding field to be explicit.  For consistency, I've added the padding to the uapi header as well.

6by9 commented on drivers/staging/vc04_services/vchiq-mmal/mmal-parameters.h:
```diff
@@ -780,7 +836,101 @@ struct mmal_parameter_camera_info {
+struct mmal_parameter_gamma {
+	u32 enabled;
+	u16 X[MMAL_NUM_GAMMA_PTS];
+	u16 Y[MMAL_NUM_GAMMA_PTS];
```
Captials.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+ */
+struct bcm2835_isp_node {
+	int vfl_dir;
+	int id;
```
unsigned int?

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	}
+
+	ret = vchiq_mmal_port_set_format(dev->mmal_instance,
+					 &dev->component->output[id]);
```
If this is called for the input node, the id = 0,  but is the input node.
Why are we setting up 2 ports here anyway? Something feels rather odd.

naushir replied:
This is a bug.  Should not be setting up the second port here!  Fixed.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	struct vchiq_mmal_port *port;
+	unsigned int size;
+
+	if (setup_mmal_component(node))
```
Are we really only setting up one port? Might be worth renaming the function accordingly.

naushir replied:
Done.  Called it setup_mmal_port().

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	struct bcm2835_isp_node *node = vb2_get_drv_priv(q);
+	struct bcm2835_isp_dev *dev = node_get_dev(node);
+	struct vchiq_mmal_port *port = get_port_data(node);
+	int ret, i;
```
unsigned int i

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+	/*
+	 * Setup the node type-specific params.
+	 *
+	 * Only the output node can set controls and crop windows. However,
```
Captialise output to denote V4L2 OUTPUT, rather than MMAL output port.

naushir replied:
Done.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1705 @@
+					MEDIA_PAD_FL_SINK : MEDIA_PAD_FL_SOURCE;
+	}
+
+	i = 0;
```
Why reset i to 0 here? It's not used until the ```for``` loop below.

naushir replied:
This was leftover from a previous change.  Removed.

6by9 reviewed (commented):
Some of these are minor nit picks, but seeing as I think we need another revision (feel free to claim otherwise) it'd be nice to see them fixed.

naushir reviewed (commented):
Update incoming.

6by9 commented on drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c:
```diff
@@ -0,0 +1,1628 @@
+	struct vchiq_mmal_port *port = get_port_data(node);
+	struct bcm2835_isp_dev *dev = node_get_dev(node);
+	unsigned int enable = 1;
+	int id = node->id;
```
id is now only used in a log line. Remove it.

naushir replied:
Done.

6by9 reviewed (commented):
Last comment as I've had enough of reading through this so many times!

naushir commented:
Thanks @6by9 :)

With that fixed, and if there are no other objections, could @popcornmix or @pelwell please merge this?

pelwell commented on drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c:
```diff
@@ -3281,6 +3282,7 @@ static int vchiq_probe(struct platform_device *pdev)
 	bcm2835_codec = vchiq_register_child(pdev, "bcm2835-codec");
 	bcm2835_camera = vchiq_register_child(pdev, "bcm2835-camera");
 	bcm2835_audio = vchiq_register_child(pdev, "bcm2835_audio");
+	bcm2835_isp = vchiq_register_child(pdev, "bcm2835-isp");
```
I object to having to enumerate the clients in code, but that's not your problem.

naushir commented:
Sorry, had to push a new revision - missed a silly typo that I had fixed locally.

pelwell commented on Documentation/media/uapi/v4l/meta-formats.rst:
```diff
@@ -19,6 +19,7 @@ These formats are used for the :ref:`metadata` interface only.
 .. toctree::
```
There's still a typo in the subject of this commit.

naushir replied:
Noted and will fix.

naushir reviewed (commented):
Let me know when you have gone though everything and I will push an update with all fixes.

pelwell commented:
I don't see anything else. It compiles cleanly, so if you push an update after final testing I'll merge it.

naushir commented:
Update pushed with typo in the commit message fixed.
